FIX Allow for KMean's attributes to be readonly

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes https://github.com/scikit-learn/scikit-learn/issues/24253


#### What does this implement/fix? Explain your changes.
This PR allows for the attributes for KMeans to be read-only. When `update_centers=False`, `lloyd_iter_chunked_dense` does not actually update `centers_new`. This is the case when `KMeans.predict` ends up calling `_labels_inertia`:

https://github.com/scikit-learn/scikit-learn/blob/feaf3829969b640594a145029e3adcffd2db0999/sklearn/cluster/_kmeans.py#L808

#### Any other comments?
Side note: I think the new test in this PR for read only attributes can become a common test.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan committed:
TST Adds failing test

thomasjpfan committed:
FIX Allow for cluster center to be readonly

thomasjpfan committed:
DOC Adds whats new entry

thomasjpfan committed:
CLN Less diff

thomasjpfan committed:
DOC Adds whats new number

ogrisel reviewed (commented):
LGTM, thanks for the fix.

Can you please just update the docstrings of `lloyd_iter_chunked_dense` and `lloyd_iter_chunked_sparse` to make it explicit that `centers_new` and `centers_squared_norms` can be left to `None` if `update_centers is False`?

ogrisel reviewed (approved):
I meant to give my approval instead of just a comment in the above review.

ogrisel commented:
@jjerphan @jeremiedbb or @fcharras might be interested in giving this PR a second look.

jjerphan commented on sklearn/cluster/tests/test_k_means.py:
```diff
@@ -1213,3 +1214,23 @@ def test_feature_names_out(Klass, method):
+    kmeans.labels_ = create_memmap_backed_data(kmeans.labels_)
+
+    y_pred2 = kmeans.predict(X)
+    assert_allclose(y_pred1, y_pred2)
```
I think they must be equal. WDYT?
```suggestion
    assert_array_equal(y_pred1, y_pred2)
```

jjerphan commented on sklearn/cluster/_kmeans.py:
```diff
@@ -795,13 +794,14 @@ def _labels_inertia(
         centers,
-        centers,
-        weight_in_clusters,
+        None,
+        None,
         labels,
         center_shift,
         n_threads,
```
I think this makes the `None`'d parameters explicit.
```suggestion
        centers_new=None,
        weight_in_clusters=None,
        labels=labels,
        center_shift=center_shift,
        n_threads=n_threads,
```

jjerphan commented on doc/whats_new/v1.2.rst:
```diff
@@ -138,6 +138,9 @@ Changelog
   :class:`cluster.AgglomerativeClustering` and will be renamed to `metric` in v1.4.
   :pr:`23470` by :user:`Meekail Zain <micky774>`.
 
+- |Fix| :class:`cluster.KMeans` support readonly attributes when predicting.
+  :pr:`24258` by `Thomas Fan`_
```
```suggestion
- |Fix| :class:`cluster.KMeans` now supports readonly attributes when predicting.
  :pr:`24258` by `Thomas Fan`_
```

jjerphan reviewed (approved):
LGTM. Thank you for fixing this issue.

> Side note: I think the new test in this PR for read only attributes can become a common test.

I am OK making the test a common test. IMO, this can be done as part of this PR.

Side note: we recently discovered that some part of the routines can be simplified, see: https://github.com/scikit-learn/scikit-learn/issues/24262

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jeremiedbb reviewed (approved):
LGTM as well after this comment https://github.com/scikit-learn/scikit-learn/pull/24258#pullrequestreview-1084983634 is addressed

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into kmeans_fix_predict_changing_state

thomasjpfan committed:
DOC Clarify conditional for paramters that can be None

thomasjpfan committed:
DOC Remove unneeded docstring parameter

thomasjpfan commented:
Updated PR with:

- [`bf3999c` (#24258)](https://github.com/scikit-learn/scikit-learn/pull/24258/commits/bf3999c7707f42395b9b8c051d9369c77dd1bb7c): Addresses comment in https://github.com/scikit-learn/scikit-learn/pull/24258#pullrequestreview-1084983634
- [`07359d8` (#24258)](https://github.com/scikit-learn/scikit-learn/pull/24258/commits/07359d8bf46cea0c9523fdd9b0c51332f61bf4a8): Removed the docstring for `centers_squared_norms` since it is not actually passed in.
